File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/engine/discovery/PhoneInvite.kt
package org.wyrdsekai.app.engine.discovery

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi

/**
 * /P5 — parse a `wyrdphone://` connection invite.
 *
 * Minted by `wyrd phone invite` (relay /phone-invite endpoint):
 *   wyrdphone://host[:port]/<base64url-JSON>
 * payload `{ v, kind:"phone", relays:[{ws_url, nats_user, nats_password,
 * fp?, ca_fp?}], household_id, zone_id, minted_at }`.
 *
 * `relays` is an ORDERED failover list (one entry today). `fp`/`ca_fp`
 * appear only for self-signed relays and seed the HouseholdTrustStore
 * TOFU pin; ACME relays carry no pin material (system trust applies).
 *
 * Pure parsing only — callers persist into [SavedHouseholdConfig] and
 * the trust store. Throws [IllegalArgumentException] with a readable
 * message on malformed input; the connect screen surfaces it verbatim.
 */
data class PhoneInvite(
    val relays: List<Relay>,
    val householdId: String?,
    /** Zone hint — lets the client skip the wyrd.discover.zone round trip. */
    val zoneId: String?,
    val mintedAt: Long?,
) {
    data class Relay(
        val wsUrl: String,
        val natsUser: String,
        val natsPassword: String,
        /** Relay leaf-cert SHA-256 (colon-hex) — self-signed relays only. */
        val fp: String?,
        /** Household CA SHA-256 (colon-hex) — self-signed relays only. */
        val caFp: String?,
    )

    companion object {
        private const val SCHEME = "wyrdphone://"

        fun isPhoneInviteUrl(text: String): Boolean =
            text.trim().lowercase().startsWith(SCHEME)

        @OptIn(ExperimentalEncodingApi::class)
        fun parse(url: String): PhoneInvite {
            val trimmed = url.trim()
            require(isPhoneInviteUrl(trimmed)) { "Not a wyrdphone:// invite URL" }
            val rest = trimmed.substring(SCHEME.length)
            val slash = rest.indexOf('/')
            require(slash > 0 && slash < rest.length - 1) {
                "Invite URL is missing its payload"
            }
            val json = try {
                val decoded = Base64.UrlSafe.withPadding(Base64.PaddingOption.ABSENT_OPTIONAL)
                    .decode(rest.substring(slash + 1))
                Json.parseToJsonElement(decoded.decodeToString()).jsonObject
            } catch (e: Exception) {
                throw IllegalArgumentException(
                    "Invite payload is not valid (re-copy the full URL)", e)
            }

            val kind = json["kind"]?.jsonPrimitive?.content
            require(kind == "phone") { "Not a phone invite (kind=${kind ?: "missing"})" }
            val relayArray = json["relays"]?.jsonArray
            require(!relayArray.isNullOrEmpty()) { "Invite carries no relays" }

            val relays = relayArray.mapIndexed { i, el ->
                val o = el.jsonObject
                val wsUrl = o["ws_url"]?.jsonPrimitive?.content
                val user = o["nats_user"]?.jsonPrimitive?.content
                val password = o["nats_password"]?.jsonPrimitive?.content
                require(!wsUrl.isNullOrEmpty() && !user.isNullOrEmpty()
                        && !password.isNullOrEmpty()) {
                    "Relay entry ${i + 1} is incomplete"
                }
                Relay(
                    wsUrl = wsUrl,
                    natsUser = user,
                    natsPassword = password,
                    fp = o["fp"]?.jsonPrimitive?.content,
                    caFp = o["ca_fp"]?.jsonPrimitive?.content,
                )
            }
            return PhoneInvite(
                relays = relays,
                householdId = json["household_id"]?.jsonPrimitive?.content.unspecifiedToNull(),
                zoneId = json["zone_id"]?.jsonPrimitive?.content.unspecifiedToNull(),
                mintedAt = json["minted_at"]?.jsonPrimitive?.longOrNull,
            )
        }

        private fun String?.unspecifiedToNull(): String? =
            if (isNullOrEmpty() || this == "unspecified") null else this
    }
}

File: clients/rn/src/network/phoneInvite.ts
/**
 * /P5 — parse a `wyrdphone://` connection invite.
 *
 * Minted by `wyrd phone invite` (relay /phone-invite endpoint). Shape:
 *   wyrdphone://host[:port]/<base64url-JSON>
 * where the payload is
 *   { v: 1, kind: "phone",
 *     relays: [{ ws_url, nats_user, nats_password, fp?, ca_fp? }, ...],
 *     household_id, zone_id, minted_at }
 *
 * `relays` is an ORDERED failover list (one entry today). `fp`/`ca_fp`
 * are present only for self-signed relays — they seed the TOFU pin the
 * native trust layer (#705) otherwise learns on first connect. ACME
 * relays carry no pin material; system trust applies.
 *
 * Pure + side-effect free so it unit-tests without an emulator. The
 * caller (ConnectScreen paste path, QR scan later) persists the fields.
 */

export interface PhoneInviteRelay {
  wsUrl: string;
  natsUser: string;
  natsPassword: string;
  /** Relay leaf-cert SHA-256, colon-hex — TOFU pin seed (self-signed only). */
  fp?: string;
  /** Household CA SHA-256, colon-hex (self-signed only). */
  caFp?: string;
}

export interface PhoneInvite {
  relays: PhoneInviteRelay[];
  householdId?: string;
  /** Zone hint — lets the client skip the wyrd.discover.zone round trip. */
  zoneId?: string;
  mintedAt?: number;
}

export function isPhoneInviteUrl(text: string): boolean {
  return text.trim().toLowerCase().startsWith('wyrdphone://');
}

/**
 * Parse an invite URL. Throws Error with a human-readable message on any
 * malformation — the connect screen surfaces it verbatim.
 */
export function parsePhoneInvite(url: string): PhoneInvite {
  const trimmed = url.trim();
  if (!isPhoneInviteUrl(trimmed)) {
    throw new Error('Not a wyrdphone:// invite URL');
  }
  const rest = trimmed.substring('wyrdphone://'.length);
  const slash = rest.indexOf('/');
  if (slash <= 0 || slash === rest.length - 1) {
    throw new Error('Invite URL is missing its payload');
  }
  const payloadB64 = rest.substring(slash + 1);

  let payload: any;
  try {
    payload = JSON.parse(base64UrlDecode(payloadB64));
  } catch (e) {
    throw new Error('Invite payload is not valid (re-copy the full URL)');
  }
  if (payload?.kind !== 'phone') {
    throw new Error(`Not a phone invite (kind=${payload?.kind ?? 'missing'})`);
  }
  if (!Array.isArray(payload.relays) || payload.relays.length === 0) {
    throw new Error('Invite carries no relays');
  }

  const relays: PhoneInviteRelay[] = payload.relays.map((r: any, i: number) => {
    if (!r?.ws_url || !r?.nats_user || !r?.nats_password) {
      throw new Error(`Relay entry ${i + 1} is incomplete`);
    }
    return {
      wsUrl: String(r.ws_url),
      natsUser: String(r.nats_user),
      natsPassword: String(r.nats_password),
      fp: r.fp ? String(r.fp) : undefined,
      caFp: r.ca_fp ? String(r.ca_fp) : undefined,
    };
  });

  return {
    relays,
    householdId: unspecifiedToUndefined(payload.household_id),
    zoneId: unspecifiedToUndefined(payload.zone_id),
    mintedAt: typeof payload.minted_at === 'number' ? payload.minted_at : undefined,
  };
}

function unspecifiedToUndefined(v: unknown): string | undefined {
  if (typeof v !== 'string' || v.length === 0 || v === 'unspecified') return undefined;
  return v;
}

function base64UrlDecode(b64url: string): string {
  let b64 = b64url.replace(/-/g, '+').replace(/_/g, '/');
  while (b64.length % 4 !== 0) b64 += '=';
  // global.atob exists in RN Hermes; Buffer covers Jest/node.
  if (typeof atob === 'function') {
    // atob yields latin1; re-decode as UTF-8 for non-ASCII zone names.
    const latin1 = atob(b64);
    const bytes = new Uint8Array(latin1.length);
    for (let i = 0; i < latin1.length; i++) bytes[i] = latin1.charCodeAt(i);
    return utf8Decode(bytes);
  }
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  return require('buffer').Buffer.from(b64, 'base64').toString('utf8');
}

function utf8Decode(bytes: Uint8Array): string {
  if (typeof TextDecoder !== 'undefined') {
    return new TextDecoder('utf-8').decode(bytes);
  }
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  return require('buffer').Buffer.from(bytes).toString('utf8');
}
